src/python/pants/rules/core/repl.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+from pathlib import PurePath
+
+from pants.base.build_root import BuildRoot
+from pants.engine.addressable import Addresses
+from pants.engine.console import Console
+from pants.engine.fs import Digest, DirectoryToMaterialize, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.interactive_runner import InteractiveProcessRequest, InteractiveRunner
+from pants.engine.objects import union
+from pants.engine.rules import goal_rule
+from pants.engine.selectors import Get
+from pants.option.global_options import GlobalOptions
+from pants.util.contextutil import temporary_dir
+
+
+class ReplOptions(GoalSubsystem):
+  """Opens a REPL."""
+  name = 'repl2'
+
+
+class Repl(Goal):
+  subsystem_cls = ReplOptions
+
+
+@union
+class ReplImplementation:
```

Comment:
Nothing is actually using this union, so this is likely to still cause "isolated backend" test failures.

I think that to actually bring the union into play , you'll need to construct an instance of the union member. That would look like taking `UnionMembership`, finding "a matching repl instance" (because this hasn't been fully generalized, you could do something like "taking the first one"), constructing an instance of it (`PythonRepl` in this case), which will allow you to remove the `targets_to_python_repl` rule.

Constructing an instance of a union member might look like:
```
# TODO: We're only expecting one of these right now: to further generalize this, we'll want
# to finish #4535.
repl_impl = next(union_membership.union_rules[ReplImplementation])
repl_binary = await Get[ReplBinary](ReplImplementation, repl_impl(addresses))
```